google-cloud-spanner/src/main/java/com/google/cloud/spanner/MutationGroup.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import com.google.common.base.Preconditions;
+import com.google.spanner.v1.BatchWriteRequest;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Represents a group of Cloud Spanner mutations to be committed together.
+ */
+public class MutationGroup {
+  private final List<Mutation> mutations;
+
+  private MutationGroup(List<Mutation> mutations) {
+    this.mutations = mutations;
+  }
+
+  public static MutationGroup of(Mutation... mutations) {
+    Preconditions.checkArgument(mutations.length > 0, "Should pass in at least one mutation.");
+    return new MutationGroup(Arrays.asList(mutations));
+  }
+
+  public List<Mutation> getMutations() {
+    return mutations;
+  }
+
+  static BatchWriteRequest.MutationGroup toProto(final MutationGroup mutationGroup) {
+    List<com.google.spanner.v1.Mutation> mutationsProto = new ArrayList<>();
+    Mutation.toProto(mutationGroup.getMutations(), mutationsProto);
+    return BatchWriteRequest.MutationGroup.newBuilder().addAllMutations(mutationsProto).build();
```

Comment:
Maybe it's a bit too much of pre-optimization, but this seems to make an unnecessary copy of the list of mutations:
1. First it creates an ArrayList and then `Mutation.toProto(...)` adds each mutation to that list.
2. Then after that we copy that ArrayList to the `BatchWriteRequest`.
 
I think it would be more efficient (and maybe also simpler) to just iterate through the list of mutations, convert one and then directly add them to the BatchWriteRequest.Builder.

Reply:
`Mutation.toProto()` goes through the list of mutations and has some additional logic to convert it into a list of mutation protos (which may reduce the size of the list also). That logic is not trivial. I think, simply iterating through the list of mutations, converting into mutation protos one by one, and adding to the BatchWriteRequest builder is not gonna directly work.

Reply:
Sounds reasonable.